File: packs/data/aws-dynamodb/scaffold/wiring/python/app/infrastructure/repositories/memory_note_repository.py
from app.domain.entities.note import Note
from app.domain.interfaces.note_repository import NoteRepository


class MemoryNoteRepository(NoteRepository):
    """In memory implementation, copied from the python-fastapi backend.

    It is the column of the contract suite that always runs and the behavior the DynamoDB adapter
    must match exactly.
    """

    def __init__(self) -> None:
        self._notes: dict[str, Note] = {}

    async def add(self, note: Note) -> Note:
        self._notes[note.id] = note
        return note

    async def get(self, note_id: str) -> Note | None:
        return self._notes.get(note_id)

    async def list_all(self) -> list[Note]:
        return list(self._notes.values())

File: packs/data/aws-dynamodb/scaffold/wiring/typescript/src/infrastructure/memory-note-repository.ts
import { type Note, noteSchema } from "../domain/note.js";
import type { NoteRepository } from "../domain/note-repository.js";

/**
 * The in memory reference implementation, copied from the typescript-node backend. It is the column
 * of the contract suite that always runs, and the behavior the DynamoDB adapter must match exactly.
 * The noteSchema.parse call is the boundary where raw storage data becomes an entity or not at all.
 */
export class MemoryNoteRepository implements NoteRepository {
  readonly #notes = new Map<string, Note>();

  add(note: Note): Promise<Note> {
    const stored = noteSchema.parse(note);
    this.#notes.set(stored.id, stored);
    return Promise.resolve(stored);
  }

  get(id: string): Promise<Note | null> {
    const found = this.#notes.get(id);
    return Promise.resolve(found === undefined ? null : noteSchema.parse(found));
  }

  listAll(): Promise<Note[]> {
    return Promise.resolve([...this.#notes.values()].map((note) => noteSchema.parse(note)));
  }
}
